chain/chain/src/chain.rs

```diff
@@ -4461,24 +4461,27 @@ impl Chain {
         shard_layout: &ShardLayout,
     ) -> Vec<CryptoHash> {
         // Using a BTreeMap instead of HashMap to enable in order iteration
-        // below.
+        // below. It's important here to use the ShardIndexes, rather than
+        // ShardIds since the latter are not guaranteed to be in order.
         //
         // Pre-populating because even if there are no receipts for a shard, we
         // need an empty vector for it.
         let mut result: BTreeMap<_, _> =
-            shard_layout.shard_ids().map(|shard_id| (shard_id, vec![])).collect();
+            shard_layout.shard_ids().enumerate().map(|shard_info| (shard_info, vec![])).collect();
```

Comment:
nit:

so here with the `enumerate()` we are relying on the fact that the shard indexes of this shard layout match the order of shard IDs returned by `shard_ids()`. Maybe it would be more future proof to not assume that? And even if we dont expect that ever to change/are ok with that assumption being in the code, the code might get simpler without this part, since it seems to me like we don't really need a `BTreeMap` here really. Since the `ShardIndex` is meant to represent an index into an array of length `num_shards()`, what if instead here we just make `result` a `Vec<Vec<Receipt>>` indexed by `ShardIndex` (maybe by adding some helper function to `ShardLayout` or making `num_shards()` `pub` or something?) and then in the loop below we just push the receipt to `result[shard_index]` and then in the return statement just do `result.into_iter().enumerate().map(|(shard_index, receipts)| { let shard_id = shard_layout.get_shard_id(shard_index).unwrap(); ... }`

Reply:
Agree with Marcelo. It might be a good idea to push the enumerate method inside shard_layout and it returns ShardInfo, i.e. (shard_index, shard_id)

Reply:
Yeah, good point, I'll follow up on that!